File: apps/mobile/targets/positions-widget/Sources/Utilities/PriceMath.swift
import Foundation

enum PriceMath {
  static func tickToPrice(tick: Int, decimalsDelta: Int) -> Double {
    let infinityThreshold = 800_000
    if tick >= infinityThreshold { return .infinity }
    if tick <= -infinityThreshold { return 0 }
    return exp(Double(tick) * log(1.0001) + Double(decimalsDelta) * log(10))
  }

  static func format(_ value: Double) -> String {
    guard value.isFinite else { return "∞" }
    if value <= 0 { return "0" }
    if value >= 1e15 { return "∞" }
    let abs = Swift.abs(value)
    if abs >= 1e6 { return compact(value) }
    if abs >= 1000 { return decimal(value, fractionDigits: 0) }
    if abs >= 1 { return decimal(value, fractionDigits: 2) }
    if abs >= 0.01 { return decimal(value, fractionDigits: 4) }
    if abs >= 1e-6 { return decimal(value, fractionDigits: 6) }
    return String(format: "%.2e", value)
  }

  private static func compact(_ value: Double) -> String {
    let abs = Swift.abs(value)
    if abs >= 1e12 { return "\(decimal(value / 1e12, fractionDigits: 2))T" }
    if abs >= 1e9 { return "\(decimal(value / 1e9, fractionDigits: 2))B" }
    if abs >= 1e6 { return "\(decimal(value / 1e6, fractionDigits: 2))M" }
    return decimal(value, fractionDigits: 0)
  }

  private static func decimal(_ value: Double, fractionDigits: Int) -> String {
    let formatter = NumberFormatter()
    formatter.numberStyle = .decimal
    formatter.minimumFractionDigits = 0
    formatter.maximumFractionDigits = fractionDigits
    formatter.usesGroupingSeparator = false
    return formatter.string(from: NSNumber(value: value)) ?? "\(value)"
  }
}

File: apps/mobile/modules/widget-bridge/android/src/main/java/expo/modules/widgetbridge/widget/util/PriceMath.kt
package expo.modules.widgetbridge.widget.util

import java.text.NumberFormat
import java.util.Locale
import kotlin.math.abs
import kotlin.math.exp
import kotlin.math.ln

object PriceMath {
  private const val INFINITY_THRESHOLD = 800_000

  fun tickToPrice(tick: Int, decimalsDelta: Int): Double {
    if (tick >= INFINITY_THRESHOLD) return Double.POSITIVE_INFINITY
    if (tick <= -INFINITY_THRESHOLD) return 0.0
    return exp(tick * ln(1.0001) + decimalsDelta * ln(10.0))
  }

  fun format(value: Double): String {
    if (!value.isFinite()) return "∞"
    if (value <= 0) return "0"
    if (value >= 1e15) return "∞"
    val a = abs(value)
    return when {
      a >= 1e6 -> compact(value)
      a >= 1000 -> decimal(value, 0)
      a >= 1 -> decimal(value, 2)
      a >= 0.01 -> decimal(value, 4)
      a >= 1e-6 -> decimal(value, 6)
      else -> String.format(Locale.US, "%.2e", value)
    }
  }

  private fun compact(value: Double): String {
    val a = abs(value)
    return when {
      a >= 1e12 -> "${decimal(value / 1e12, 2)}T"
      a >= 1e9 -> "${decimal(value / 1e9, 2)}B"
      a >= 1e6 -> "${decimal(value / 1e6, 2)}M"
      else -> decimal(value, 0)
    }
  }

  private fun decimal(value: Double, fractionDigits: Int): String {
    val f = NumberFormat.getInstance(Locale.US).apply {
      minimumFractionDigits = 0
      maximumFractionDigits = fractionDigits
      isGroupingUsed = false
    }
    return f.format(value)
  }
}
